### PR TITLE
Breakdown UI for Funnels

### PR DESCRIPTION
**Pull request by samwinslow**
## Changes

Adds breakdowns to the Funnel Steps UI. Thanks to @liyiy for the help with testing this on Clickhouse.

Screen Shot 2021-07-12 at 8 45 05 AM

Hovering on a bar reveals more context about the breakdown value of that particular step:

Screen Shot 2021-07-12 at 8 45 16 AM

The full hover context menu (shown when dropoffs >0) looks like this:

Screen Shot 2021-07-12 at 8 45 27 AM

Parts of this PR are somewhat rough, though I'd like to get this in as soon as possible in order to test it with a user today. The "dragons" here that I will refactor in the near future are:

- [x] Splitting up `FunnelBarGraph.tsx` into many component files
- [ ] `decideLabelPosition()` which determines whether and where to place the percent indicator in the bar (it's buggy)
- [x] The type definition of an API response with breakdown values — currently `FunnelResultWithBreakdown.result` is `FunnelStep[][]`, but you will see a strange cast to `any` since the return type of `pollFunnel` has not been fully updated

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests
- [x] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [x] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by liyiy** [comment]
This shows up when I hover over the bar without a breakdown applied

Screen Shot 2021-07-12 at 11 12 32 AM

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (samwinslow) [inline frontend/src/scenes/funnels/funnelLogic.ts:160]: Tricky: `pollFunnel` would ideally return a `FunnelResult` or `FunnelResultWithBreakdown` based on whether its input params include `breakdown`. But I don't want to handle the possibility of `FunnelResultWithBreakdown` and `FunnelResult` everywhere, which is why the `aggregateBreakdownResult` function comes in to reshape the data.

File: frontend/src/scenes/insights/InsightTooltip/index.tsx
Comment (samwinslow) [inline frontend/src/scenes/insights/InsightTooltip/index.tsx:29]: `bodyLines` is in here for its similarity to the charting library's built-in tooltips, but `children` is easier to use in other React components

File: frontend/src/types.ts
Comment (samwinslow) [inline frontend/src/types.ts:693]: By making this a recursive type, we can construct an array where the the top-level objects contain count and step metadata for all breakdown values, and the children describe subsets of that step.

File: frontend/src/types.ts
Comment (samwinslow) [inline frontend/src/types.ts:705]: The type of this API response is flattened by `aggregateBreakdownResult` into a better type.

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (samwinslow) [inline frontend/src/scenes/funnels/funnelLogic.ts:33]: In this case, `step.breakdown` is a string

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (samwinslow) [inline frontend/src/scenes/funnels/funnelLogic.ts:37]: We could concatenate and de-dupe the UUIDs from breakdown values here, but it sounds expensive and Li's person modal fetches the data again anyway.

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (samwinslow) [inline frontend/src/scenes/funnels/funnelLogic.ts:34]: Use of `reduce` here is to compensate for the fact that the API response does not include aggregated / hierarchical data. Have talked with @EDsCODE about changing the shape of the response value — I can possibly even take that on myself in the future

**Review by mariusandra — COMMENTED** [review]
Hey, I was not able to get any funnels showing locally, not even on `master` ([internal slack link](https://posthog.slack.com/archives/C019RAX2XBN/p1626259504000800)). Thus I didn't really see the breakdown :(. 

If anyone else has better luck with getting the API to return data, please defer to their review for that part. I didn't see any obvious issues with that code.

You left a bunch of comments below in the PR to comment your code, but these will be lost when the PR is merged, yet the code won't be. 🤔 If only there was a way to `// comment` code more permanently 😆 . Joking aside, if you comment your code, it should be to help people who will see this code later, not to "pass the review". So put them where they are most relevant.

The main comment I have for the code is to find a way to remove the `any`s.

**Comment by mariusandra** [comment]
So.. finally got my CH fixed. In addition to the comments above, I just found this small issue with very long breakdown values: 
image

Otherwise this seems to work great!

**Comment by samwinslow** [comment]
@alexkim205 I standardized the types somewhat by using generics — does this look OK to you?

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:69]: Is this a `&` or a `|` actually?

File: frontend/src/types.ts
Comment (mariusandra) [inline frontend/src/types.ts:699]: This comment is not very useful in that it doesn't answer the question "why?". Why do we need this MappedFunnelStep? Why is it "more useful"? 

Could there be a better name that describes the difference between FunnelStep and MappedFunnelStep? The prefix "Mapped" on its own doesn't really say anything. How is the data different? I'd even say something very verbose like `FunnelStepWithAggregatedBrekdown` would fit better.

Ideally I would avoid this custom type altogether.

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:70]: Could you clarify? Or better yet, document with types? Function overloading is possible with TypeScript.

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:163]: Could you type it accordingly then? Removing a type, and adding `any` with a 🤷‍♂️ comment is not really the solution :).

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:189]: I think we're doing things in a wrong order here. 

Ideally, a `loader` should just have one job: loading data from an API endpoint... or two. Here the loader also takes on the role of transforming this data and doing various other things.

If you want to transform loaded data into some other format, you should use a selector or a combination of selectors. E.g. `results` -> `filteredResults` -> `sortedResults`. 

Could some similar pattern work here as well? E.g. you just use a selector on `steps` that aggregates them? Then all places where you need this breakdown data just read from that selector.

Could this work?

**Review by mariusandra — CHANGES_REQUESTED** [review]
This TypeScript story still feels off...

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (samwinslow) [reply frontend/src/scenes/funnels/funnelLogic.ts:189]: Yep, selectors are definitely the way to go. What's funky as well is that the second call to `pollFunnel` to fetch histogram bins data does not return to the enclosing scope, so that probably belongs in its own listener/loader.

**Comment by samwinslow** [comment]
I will try my best to explain my reasoning here.

The `FunnelStepWithNestedBreakdown` type has breakdown data included in an array, but the top level steps look just like `FunnelStep`s, and can be iterated over in just the same way, which avoids having to handle `results` being either a 1- or 2-dimensional array.

`results` is a 2-dimensional array when `breakdown` is passed as a query parameter to the API. I have already told Core Analytics that I would prefer the return type to always be a 1-dimensional array, with the `breakdown` property set on each step to include data for each breakdown.

In almost every place where `FunnelStep` or `FunnelStepWithNestedBreakdown` is used, the type of `breakdown` doesn't even matter because it is not accessed unless we KNOW the steps have breakdown values, because we have the params that were sent to the API.

Yes, `pollFunnel` should have overloads for different types, but when I tried doing this, TS complained that the return type could not be inferred at compile time. So, I have made `pollFunnel` generic. In pursuit of removing a couple instances of `any`, I have now filled this code with type assertions.

Additionally, there was code in funnelLogic already approved via #5050 either without types or with incorrect types, including those for `timeConversionBins` and chaining `Promise<FunnelResult & FunnelsTimeConversionResult>`.

I am legitimately trying as hard as I can to get this right while also just getting it done.

**Comment by samwinslow** [comment]
1. Apparently can't iterate over `steps` (in `stepsWithCount`) if the type is `(FunnelStep | FunnelStepWithNestedBreakdown)[]` ??
2. `getReferenceStep` is literally just a fancy array accessor. The signature looks like `function getReferenceStep<T>(steps: T[], stepReference: FunnelStepReference, index?: number): T` but upon invocation with steps as `(FunnelStep | FunnelStepWithNestedBreakdown)[]` the return type is always assumed to be `FunnelStep` ???

**Comment by mariusandra** [comment]
This is ready for a look now. I completely redid the logic for the loaders (including for the time conversion bins), refactored the types to be even saner, and fixed various bugs and inconsistencies I found.

I didn't test all possible cases, so it'd be good to have another set of eyes on this before merging it in.

**Review by paolodamico — APPROVED** [review]
one minor comment for the future but lgtm!

**Pull request merged**
